src/app/browser/qgsinbuiltdataitemproviders.cpp

```diff
@@ -495,7 +495,10 @@ void QgsLayerItemGuiProvider::deleteLayers( const QStringList &itemPaths )
       QgsMessageLog::logMessage( tr( "Item with path %1 no longer exists." ).arg( itemPath ) );
       return;
     }
-    if ( !item->deleteLayer() )
+    Q_NOWARN_DEPRECATED_PUSH
```

Comment:
This is because only gdal and ogr data items have been turned into the "new" system of using QgsDataItemGuiProvider for rename(), deleteLayer() etc simply because they had to. Other data items have been kept as they were for two reasons - 1. to make sure the "legacy" approach still works, 2. it would be even more massive code shuffle.

In the "next steps" section I have left a note that more data items should be migrated to the "new" system....

Reply:
Right... makes sense. In general I thought we had the mentality "you deprecate it, you remove it" and that it was the responsibility of anyone deprecating an API call to ensure that there's no remaining uses left of it in master (which aren't themselves there for legacy reasons). I remember a couple of years back someone (@m-kuhn ?) did a huge amount of work cleaning up a ton of deprecated uses because people had been deprecating stuff without upgrading all the code paths, and I'd hate to see us slide back into this again... That said, there's obviously a ton of work already done here, and I can understand you would want to finalise this quickly.

I trust whatever decision you make here 😄 

On that note... We've collected quite a number of deprecated methods since 3.0. I wonder when the discussion needs to begin about setting a future date for 4.0. My gut feeling is that (in a year? 2 years?) we could do a bump to 4.0, adopting the same approach as Qt is doing for 6.0 (i.e. targeting very minimal breakage). But we should probably start a public discussion soon, and set some roadmap... even if it's years in the future!

Reply:
Yeah it is a tough decision to do :-) On one hand I agree with the "you deprecate it, you remove it" mantra, but here that means adding maybe ~1k of changed lines of code :-/ And there are already conflicts starting to pile so I would rather get this merged ASAP and promise it will get addressed in on of the followup PRs. In any case, we need to keep these NOWARN bits during 3.x to avoid API breaks.

And yeah - soon it will be time to think about 4.0 - although I am sure that will be highly controversial topic again :-)

Reply:
Ok, I think you've definitely earned the trust to deviate from any usual policy!

+1 to merge asap (pending other comments) also. 

> And yeah - soon it will be time to think about 4.0 - although I am sure that will be highly controversial topic again :-)

That's why I'd love for the conversation to at least start, even if the outcome is "4.0 in 4 years". There's absolutely nothing gained from shying away from tricky decisions!

Reply:
Thanks for the trust :relaxed:

... *rolling up my sleeves again* ...

Reply:
Just to give my 2 cents: I agree with the discussion above regarding quick merge (on trust basis) and the required discussion towards QGIS 4, be it whenever :+1: